### PR TITLE
feat: track cash transfers as asset transaction type

### LINKED ISSUE #134: Track cash transfers as asset transaction type
**Extend AssetTransactionType**
```graphql
enum AssetTransactionType {
  CREATED
  BORROWED
  REPAID
  CLOSED
  CASH_DEPOSIT
  CASH_WITHDRAWAL
  CASH_TRANSFER // to/from onchain reserve 
}
```

**Create `OnchainCash` asset**
When pool is created, create an asset by default with type = `OnchainCash`.

**If there is a transfer debt from a non-cash asset to an `OffchainCash` asset**
- Track repayment amount
- Create `PrincipalRepayment` transaction for the non-cash asset, with `fromAsset = OffchainCash asset` and `amount = repaidAmount`
- Create `CashDeposit` transaction type for the cash asset, with `toAsset = non-cash asset` and `amount = borrowAmount`

=> this will show up as a principal repayment in the Cashflows chart

**If there is a transfer debt from an `OffchainCash` asset to non-cash asset**
- Track borrowed/financed amount
- Create `Purchase` transaction for the non-cash asset, with `fromAsset = OffchainCash asset` and `amount = borrowAmount`
- Create `CashWithdrawal` transaction for the cash asset, with `toAsset = non-cash asset` and `amount = repaidAmount`

=> this will show up as a purchase in the Cashflows chart

**If there is a purchase of an `OffchainCash` asset**
- Don't track any financing/repayment amounts
- Create a `CashTransfer` transaction for the cash asset, with `fromAsset = OnchainCash asset`

=> this won't show up in the Cashflows chart

**If there is a repayment of an `OffchainCash` asset**
- Don't track any financing/repayment amounts
- Create a `CashTransfer` transaction for the cash asset, with `toAsset = OnchainCash asset`

=> this won't show up in the Cashflows chart

### PR DESCRIPTION
**Pull request by filo87**
Fixes #134

### CODE REVIEW

**Comment by filo87** [comment]
@hieronx can you by chance review the logic?

File: src/mappings/handlers/loansHandlers.ts [hunk 1]
```diff
@@ -103,44 +103,62 @@ async function _handleLoanBorrowed(event: SubstrateEvent<LoanBorrowedEvent>): Pr
...
 
+  const epoch = await EpochService.getById(pool.id, pool.currentEpoch)
+  if (!epoch) throw new Error('Epoch not found!')
+
   // Update loan amount
   const asset = await AssetService.getById(poolId.toString(), loanId.toString())
   await asset.activate()
-  await asset.borrow(amount)
+
+  if (asset.type === AssetType.OffchainCash) {
```
[the remark this answers is not in the document]
Comment (filo87) [reply src/mappings/handlers/loansHandlers.ts:118]: and also group the transaction data all together

**Comment by filo87** [comment]
@hieronx can u re-check?

File: src/mappings/handlers/loansHandlers.ts [hunk 2]
```diff
@@ -278,86 +291,149 @@ async function _handleLoanDebtTransferred(event: SubstrateEvent<LoanDebtTransfer
...
+      assetId: fromLoanId.toString(10),
+      amount: repaidAmount,
+      interestAmount: repaidInterestAmount,
+      principalAmount: repaidPrincipalAmount,
+      unscheduledAmount: repaidUnscheduledAmount,
+      quantity: _repaidAmount.principal.isExternal ? _repaidAmount.principal.asExternal.quantity.toBigInt() : null,
+      settlementPrice: _repaidAmount.principal.isExternal
+        ? _repaidAmount.principal.asExternal.settlementPrice.toBigInt()
+        : null,
+      toAssetId: toLoanId.toString(10),
```
Comment (hieronx) [inline src/mappings/handlers/loansHandlers.ts:326]:
```suggestion
      fromAssetId: fromLoanId.toString(10),
      toAssetId: toLoanId.toString(10),
```

File: src/mappings/handlers/loansHandlers.ts [hunk 3]
```diff
@@ -278,86 +291,149 @@ async function _handleLoanDebtTransferred(event: SubstrateEvent<LoanDebtTransfer
...
+    const purchaseTransaction = await AssetTransactionService.borrowed({
+      ...txData,
+      assetId: toLoanId.toString(10),
+      amount: borrowPrincipalAmount,
+      principalAmount: borrowPrincipalAmount,
+      quantity: _repaidAmount.principal.isExternal ? _repaidAmount.principal.asExternal.quantity.toBigInt() : null,
+      settlementPrice: _repaidAmount.principal.isExternal
+        ? _repaidAmount.principal.asExternal.settlementPrice.toBigInt()
+        : null,
+      fromAssetId: fromLoanId.toString(10),
```
Comment (hieronx) [inline src/mappings/handlers/loansHandlers.ts:354]:
```suggestion
      fromAssetId: fromLoanId.toString(10),
      toAssetId: toLoanId.toString(10),
```

File: src/mappings/handlers/loansHandlers.ts [hunk 4]
```diff
@@ -278,86 +291,149 @@ async function _handleLoanDebtTransferred(event: SubstrateEvent<LoanDebtTransfer
...
+
+    await epoch.increaseRepayments(amount)
+    await epoch.save()
+
+    // principal repayment transaction
+    const principalRepayment = await AssetTransactionService.repaid({
+      ...txData,
+      assetId: fromLoanId.toString(10),
+      amount: amount,
+      toAssetId: toLoanId.toString(10),
```
Comment (hieronx) [inline src/mappings/handlers/loansHandlers.ts:410]:
```suggestion
      fromAssetId: fromLoanId.toString(10),
      toAssetId: toLoanId.toString(10),
```

File: src/mappings/handlers/loansHandlers.ts (hunk 3 above)
Comment (hieronx) [reply src/mappings/handlers/loansHandlers.ts:355]: @filo87 why was this one resolved rather than committed?

**Review by hieronx — APPROVED** [review]

**Pull request merged**